### PR TITLE
executor: fix left join on partition table generate invalid lock key

### LINKED ISSUE #28073: Invalid keys may get locked
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Execute the following SQL, then query the MVCC of an invalid key `7480000000000000005F728000000000000000`.
```sql
/* test */ drop table if exists t1, t2;
/* test */ create table t1  (c_int int, c_str varchar(40), primary key (c_int, c_str) , key(c_int)) partition by hash (c_int) partitions 4;
/* test */ create table t2  like t1 ;
/* test */ insert into t1 values (1, 'flamboyant mcclintock');
/* test */ insert into t2 select * from t1 ;
/* test */ begin;
/* test */ insert into t2 (c_int, c_str) values (2, 'romantic grothendieck');
/* test */ select * from t2 left join t1 on t1.c_int = t2.c_int for update;
/* test */ commit;
```

### 2. What did you expect to see? (Required)

The MVCC info should be empty.

### 3. What did you see instead (Required)

The key was locked unexpectedly. It seems the partition id column (used as the physical id) is actually null (thus [`row.GetInt64(offset)`](https://github.com/pingcap/tidb/blob/master/executor/executor.go#L933) returns 0) in this case.

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #28073

Problem Summary:

select * from t1 left join t2 ... for update
t2 is partition table.

In the left join case, the physical column id might be 0, 
because when a row in t1 but not in t2, NULL will be filled... 
and the physical column ID is not set in the row.

It cause the `for update` generate the invalid lock key with physical ID = 0

### What is changed and how it works?

Skip the physical ID = 0 lock key to avoid that case.

### Check List

Tests 

- [X] Unit test

By the way, I find something seems wrong:

- kv storage Scan() API seems don't scan the lock CF... (I'm not sure is it by design? there is also a ScanLock() API used by the GCWorker)
- We should use 'begin pessimistic' in the test so the lock key is written before the transaction commit. In optimistic txn mode, the lock key is buffered and written until commit... then the test code can't read that lock

- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by mjonss — APPROVED** [review]
LGTM, with a small question/suggestion (feel free to ignore)

File: executor/executor.go [hunk 1]
```diff
@@ -1025,6 +1025,15 @@ func (e *SelectLockExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
+						if physTblID == 0 {
+							// select * from t1 left join t2 on t1.c = t2.c for update
+							// The join right side might be added NULL in left join
+							// In that case, physTblID is 0, so skip adding the lock.
+							//
+							// Note, we can't distinguish whether it's the left join case,
+							// or a bug that TiKV return without correct physical ID column.
+							continue
+						}
 					}
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/executor.go:1037]: table ID should never be 0 ... that's change is more defensive but it might hide some other bugs we're not aware of.

**Review by bb7133 — APPROVED** [review]
LGTM

File: executor/executor.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjonss) [reply executor/executor.go:1038]: Thanks for the explanation!

**Pull request merged**